vi.py

```diff
@@ -0,0 +1,224 @@
+                "bn%s" % idx,
+                torch.nn.BatchNorm1d(hidden_features),
+            )
+
+        if depth == 1:
+            self.gn0 = layer(
+                in_feats=in_features, out_feats=out_features, activation=None
+            )
+
+        self.a_prior = torch.distributions.Normal(1.0, a_prior)
+
+        # middle layers
+        self.a_mu = torch.nn.Parameter(torch.tensor(1.0))
+
+        self.a_log_sigma = torch.nn.Parameter(
+            torch.tensor(a_log_sigma_init)
+        )
+
+        self.in_features = in_features
+        self.hidden_features = hidden_features
+        self.depth = depth
+        self.kl_scaling = kl_scaling
+
+
+
+    def _forward(self, g, x):
+        """ Internal forward.
+
+        Parameters
+        ----------
+        g : dgl.DGLGraph
+            Input graph.
+
+        x : torch.Tensor
+            Input features. (n_nodes, in_features)
+
+        Returns
+        -------
+        dgl.DGLGraph
+            Graph with perturbed weights annotated.
+
+        """
+        g = g.local_var()
+
+        g.apply_edges(
+            lambda edges: {
+                **{
+                    "a0": torch.distributions.Normal(
+                        self.a_mu, self.a_log_sigma,
+                        ).rsample([g.number_of_edges(), self.in_features]),
+                },
+                **{
+                    "a%s" % idx: torch.distribution.Normal(
+                        self.a_mu, self.a_log_sigma,
+                        ).rsample([g.number_of_edges(), self.hidden_features])
+                    for idx in range(1, self.depth)
+                }
+            }
+        )
+
+        for idx in range(self.depth):
+            g.edata["a"] = g.edata["a%s" % idx]
+            x = getattr(self, "gn%s" % idx)(g, x)
+            x = getattr(self, "bn%s" % idx)(x)
+            x = self.activation(x)
+
+        g.ndata["x"] = x
+        return g, x
+
+    @staticmethod
+    def condition(self, x):
+        """ Condition (static method. )
+
+        Takes a set of parameters and return a distribution.
+
+        Parameters
+        ----------
+        x : torch.Tensor (n_graphs, out_features)
+
+        Returns
+        -------
+        torch.distributions.Categorical
+
+        """
+        return torch.distributions.categorical.Categorical(logits=x)
+
+    def loss(self, g, x, y, n_samples=1, mask=None):
+        """ Training loss.
+
+        Parameters
+        ----------
+        g : dgl.DGLGraph
+            Input graph.
+
+        x : torch.Tensor
+            Input features.
+
+        y : torch.Tensor
+            Reference value.
+
+        n_samples : int
+            Number of samples to consider.
+
+        Returns
+        -------
+        torch.Tensor : loss
+
+        """
+        # initialize losses
+        losses = []
+
+        for _ in range(n_samples):
+            # get input graph
+            # x.shape = (n_graphs, out_features)
+            g, x = self._forward(g, x)
```

Comment:
g_ !

Reply:
x_ !

Reply:
if you want to simplify: pass x_ and g_ into the loss and do x, g = ....(x_, g_)